easybuild/tools/containers.py

```diff
@@ -0,0 +1,304 @@
+        else:
+            raise EasyBuildError("Unknown format of OS dependency specification encountered: %s", osdep)
+
+    # module names to load in container environment
+    mod_names = [e['ec'].name + '/' + det_full_ec_version(e['ec']) for e in easyconfigs]
+
+    # name of Singularity definition file
+    img_name = build_option('container_image_name')
+    if img_name:
+        def_file_label = os.path.splitext(img_name)[0]
+    else:
+        def_file_label = mod_names[0].replace('/', '-')
+
+    def_file = 'Singularity.%s' % def_file_label
+
+    # adding all the regions for writing the  Singularity definition file
+    content = SINGULARITY_TEMPLATE % {
+        'bootstrap': bootstrap_agent,
+        'from': bootstrap_from,
+        'install_os_deps': install_os_deps,
+        'easyconfigs': ' '.join(os.path.basename(e['spec']) for e in easyconfigs),
+        'mod_names': ' '.join(mod_names),
+    }
+    def_path = os.path.join(cont_path, def_file)
+
+    if os.path.exists(def_path):
+        if build_option('force'):
+            print_msg("WARNING: overwriting existing container recipe at %s due to --force" % def_path)
+        else:
+            raise EasyBuildError("Container recipe at %s already exists, not overwriting it without --force", def_path)
+
+    write_file(def_path, content)
+    print_msg("Singularity definition file created at %s" % def_path, log=_log)
+
+    return def_path
+
+
+def build_singularity_image(def_path):
+    """Build Singularity container image by calling out to 'singularity' (requires admin privileges!)."""
+
+    cont_path = container_path()
+    def_file = os.path.basename(def_path)
+
+    # use --imagename if specified, otherwise derive based on filename of recipe
+    img_name = build_option('container_image_name')
+    if img_name is None:
+        # definition file Singularity.<app>-<version, container name <app>-<version>.<img|simg>
+        dot_idx = def_file.find('.')
+        img_name = def_file[dot_idx+1:]
+
+    cmd_opts = ''
+
+    image_format = build_option('container_image_format')
+
+    # squashfs image format (default for Singularity)
+    if image_format in [None, CONT_IMAGE_FORMAT_SQUASHFS]:
+        img_path = os.path.join(cont_path, img_name + '.simg')
+
+    # ext3 image format, creating as writable container
+    elif image_format == CONT_IMAGE_FORMAT_EXT3:
+        img_path = os.path.join(cont_path, img_name + '.img')
+        cmd_opts = '--writeable'
+
+    # sandbox image format, creates as a directory but acts like a container
+    elif image_format == CONT_IMAGE_FORMAT_SANDBOX:
+        img_path = os.path.join(cont_path, img_name)
+        cmd_opts = '--sandbox'
+
+    else:
+        raise EasyBuildError("Unknown container image format specified for Singularity: %s" % image_format)
+
+    if os.path.exists(img_path):
+        if build_option('force'):
+            print_msg("WARNING: overwriting existing container image at %s due to --force" % img_path)
```

Comment:
@boegel

it may be a good idea when user is overwriting container with --force to remove existing container before building new container. Also note that ``singularity build --force`` is also an option but that is something I have not played around with. 

If this option is useful in future we would need to differentiate between ``eb --force`` and ``singularity build --force`` to avoid confusion. 


```
    -f|-F|--force   Force a rebootstrap of a base OS (note: this does not
                    delete what is currently in the image, just causes the core
                    to be reinstalled)
```

I am guessing singularity build -F may be useful if you want to build same container and there is update in the bootstrap image.